src/span.rs

```diff
@@ -0,0 +1,185 @@
+use std::{io, ops::Range};
+
+pub trait Span: Clone {
+    type Subspan: Span<Pos = Self::Pos>;
+    type Pos: Ord + Copy;
+
+    fn start(&self) -> Self::Pos;
+    fn end(&self) -> Self::Pos;
+    fn slice(&self, range: Range<Self::Pos>) -> Self::Subspan;
+}
+
+pub trait SpanFormatter<Span: self::Span> {
+    fn first_line(&self, span: &Span) -> WithLineNumber<Span::Subspan>;
+    fn next_line(
+        &self,
+        span: &Span,
+        subspan: &WithLineNumber<Span::Subspan>,
+    ) -> Option<WithLineNumber<Span::Subspan>>;
+    fn count_columns(&self, span: &Span, subspan: &Span::Subspan) -> usize;
+}
+
+pub trait SpanWriter<Span: crate::Span> {
+    fn write(&self, w: &mut dyn io::Write, span: &Span, subspan: &Span::Subspan) -> io::Result<()>;
+}
+
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub struct WithLineNumber<T> {
```

Comment:
Ohh this looks nice! If I understand it correctly, `WithLineNumber` allows to provide a span that's starting on the specific line number? If yes, that will help a lot with https://github.com/rust-lang/rust/issues/61809 :confetti_ball:

Reply:
That's actually very slightly more powerful than what I had, but not hard to add on top: 189c5f3; what it was before was just the current [`line_start`](https://docs.rs/annotate-snippets/0.6.1/annotate_snippets/snippet/struct.Slice.html#structfield.line_start) field of a `Slice` for `&str` only, as they don't carry offset information to derive line numbers like a slice does.

Now it works for "dumb" slices as well, so `WithLineNumber { data: span, line_num: num }` makes it such that the span used starts at the given line number. (This means for range spans that if the span formatter says it starts at line `5` and the `line_num == 5`, the first line will be reported as `9`.)